Updates after refactoring Connection constructor

abmusse opened the pull request:
## Changes to Toolkit.js

After refactoring Connection construct options were added to return an error on the user provided callback to `Connection.run()`. `Toolkit.js` was previously updated with https://github.com/IBM/nodejs-itoolkit/commit/a632dbf1c99371771e47bd8922df9c75ce49f4b0. These changes do not take into account when `returnError` is set to false. `returnError` is to false when `iConn` class is used. This is because in pre `v.1.0` version errors were not returned.  In that case run will only return XML output instead `(error, output)`.


Now with 423605983657f0bff34ca70e151535fef628e078: added in checks for when `returnError` is false Toolkit.js. Within the constructor a new Connection based on the transport options from the `conn` parameter is created when `returnError` is false. This was done to avoid worrying about setting it back to on at some point and possible race conditions if the same connection object is used elsewhere.

## Added CommandCall Class

`CommandCall` class was added to  to maintain command types (sh, cl, qsh). Please view this [issue](https://github.com/IBM/nodejs-itoolkit/issues/24#issuecomment-472197312) for reasoning. 

## Updated Deprecated.js

- Now [depd](https://www.npmjs.com/package/depd) package to display deprecation warnings
- wrap getConnection() to call getTransportOptions
- have setTimeout() print warningn no longer used

## Updated unit/functional tests

Both unit and functional tests updated to use current non deprecated
versions.

Moved existing unit/functional tests:
- Functional tests using deprecated class moved to `test/functional/deprecated`

- Unit tests using deprecated classes moved to `test/unit/deprecated`

Added ToolkitFunctional.js
- remove functional test deprecated files (iDataQueueFunctional, ...)
- consolidates all Toolkit methods into `ToolkitFunctional.js`

kadler commented:
Seems really weird to have a single `CommandCall` object which takes in varying configuration options and then having specific methods on it for the various types (ie. `qsh`, `cl`, `sh`). Why not just have three separate classes or have a single class which takes in a parameter for the type of command to call?

markdirish commented:
Probably better to have three separate classes I would think. I think I combined them into CommandCall when I had the zeal of cutting down as much of the duplicated code in this project as possible.

abmusse commented:
@kadler 

> Why not have a single class which takes in a parameter for the type of command to call

The current constructor for [CommandCall](https://github.com/IBM/nodejs-itoolkit/pull/45/commits/d2e8b571516d9da12ce5935787ff4b6476d3750f#diff-5e05cf0e6359f91b6a5859876734d7f0R40) accepts an object with the format:

```javascript
{
 command: ' '
 type: 'sh' || 'cl' || 'qsh',
 [options]: {}
}
```

command and type properties are required.

So it does a have a parameter for the type of command to call.

In `toXML` the appropriate method `(sh, qsh, cl)` based on the type specified.

Ideally, I would like to have `sh`, `qsh`, and `cl` methods to be private to the class but I don't think that's possible in JS yet. 

I would like to have had `toXML` function from the `CommandCall` to be the only public accessible method.

abmusse commented:
> Why not just have three separate classes?

Having 3 separate classes works too.

The reason I did not go that route first was to keep things less verbose.

If we went that route the classes themselves would be pretty thin for now.

Maybe something like: 


```javascript
class CLCommand {

    constructor(command, options) {
      if (!command || typeof command !== 'string') {
	   throw new Error('Command is required');
      }

      this.command == command;
      
      if (options && typeof options === 'object') {
         this.options = options;
      }
      
    }


    toXML() {
       .... code to generate and return xml
   }
}

```

The other 2 command classes would be similar

kadler commented:
> In `toXML` the appropriate method (sh, qsh, cl) based on the type specified is called.

I see now. In that case, I'd just merge those functions in to `toXML`

kadler commented:
I really hate this `iXmlNodeCmdOpen` stuff as well. It's all overly complicated and over-engineered.

abmusse commented:
> I really hate this `iXmlNodeCmdOpen` stuff as well. It's all overly complicated and over-engineered.

We should consider refactoring and simplifying `ixml.js` in a future PR.

kadler commented:
> We should consider refactoring and simplifying `ixml.js` in a future PR.

Agreed.

kadler commented on lib/CommandCall.js:
```diff
@@ -25,11 +25,17 @@ const {
   constructor(config) {
     if (!config.command || !config.type) {
-      throw new Error('Please specify a command and type (cl, qsh, sh)');
+      throw new Error('Please specify a command and type (cmd, qsh, sh)');
```
I preferred this is "cl", since it's more explicit. Could map to

abmusse replied:
Agreed `cl` is clear compared to `cmd`.

markdirish replied:
I see you reverted, but isn't the XMLService tag <cmd>? I think `cl` makes more sense, and not advocating reverting it back again, but just something to consider.

abmusse replied:
The xml service tag is `<cmd>`

abmusse commented:
I merged (sh, cl, qsh) functions into toXml(). 

Also renamed type cl -> cmd. Xml service uses cmd as the tag name.

Updated other places where type cl was being used to cmd since that name xml service uses.

abmusse commented:
Reverted back to use `cl` instead of `cmd`.

kadler commented on lib/CommandCall.js:
```diff
@@ -0,0 +1,86 @@
+        this.xml = iXmlNodeShOpen('', '', '', '', '') + this.command + iXmlNodeShClose();
+      }
+    } else if (this.type === 'cl') {
+      const xexec = (this.command.indexOf('?') > 0) ? 'rexx' : 'cmd';
```
`xexec`?

abmusse replied:
weird naming should just be exec:


```text
* <cmd [exec='cmd|system|rexx'
      *       hex='on' 
      *       before='cc1/cc2/cc3/cc4'
      *       after='cc4/cc3/cc2/cc1'
      *       error='on|off|fast'
      *       ]>values (see IBM i *CMD)</cmd>
```

markdirish replied:
`x<whatever>` is the naming convention of variables used in 0.x when creating XML. Its all throughout ixml.js, but agree that the `x` should be rooted out wherever it is found.

abmusse replied:
Sounds good will change it to `exec`

abmusse replied:
@kadler @markdirish 

renamed `xexec` -> `exec`

abmusse commented:
I will squash https://github.com/IBM/nodejs-itoolkit/pull/45/commits/e33cc8c73409bf45cb29d44842d5633898282168 and https://github.com/IBM/nodejs-itoolkit/pull/45/commits/e164691c35f7fc6a84c325de0e1fb20589882820 with https://github.com/IBM/nodejs-itoolkit/pull/45/commits/d2e8b571516d9da12ce5935787ff4b6476d3750f
